### PR TITLE
New Feature: Adding Index Metrics to Query Metrics

### PR DESCRIPTION
**Pull request by aayush3011**
Sample of the new Query Metrics. You can find the Index Metrics under Index Utilization Info Metrics.

image

## Benchmark Results

Screen Shot 2021-08-19 at 7 53 27 AM

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CompositeIndexUtilizationEntity.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package com.azure.cosmos.implementation;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/CompositeIndexUtilizationEntity.java:1]: New file license header missing.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
+package com.azure.cosmos.implementation;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java:1]: New file license header missing.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/metrics/QueryMetricsWriter.java [hunk 3]
```diff
@@ -4,9 +4,7 @@
 package com.azure.cosmos.implementation.query.metrics;
 
 import com.azure.cosmos.BridgeInternal;
-import com.azure.cosmos.implementation.QueryMetrics;
-import com.azure.cosmos.implementation.QueryPreparationTimes;
-import com.azure.cosmos.implementation.RuntimeExecutionTimes;
+import com.azure.cosmos.implementation.*;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/metrics/QueryMetricsWriter.java:7]: We don't use `*` imports, please use imports explicitly.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DefaultDocumentQueryExecutionContext.java [hunk 4]
```diff
@@ -185,7 +185,8 @@ protected PartitionKeyInternal getPartitionKeyInternal() {
                                                     tFeedResponse.getRequestCharge(),
                                                     this.fetchExecutionRangeAccumulator.getExecutionRanges(),
                                                     Arrays.asList(schedulingTimeSpanMap)),
-                                            tFeedResponse.getActivityId());
+                                            tFeedResponse.getActivityId(),
+                                        tFeedResponse.getResponseHeaders().getOrDefault(HttpConstants.HttpHeaders.INDEX_UTILIZATION, null));
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DefaultDocumentQueryExecutionContext.java:189]: why we have null in last parameter , also i am not seeing changes  for ClientSideMetrics ctor

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java [hunk 5]
```diff
@@ -483,6 +485,32 @@ public CosmosQueryRequestOptions setThresholdForDiagnosticsOnTracer(Duration thr
...
+    /**
+     * Sets indexMetricsPopulated, which is used to obtain the index metrics to understand how the query engine used existing
+     * indexes and could use potential new indexes.
+     * The results will be displayed in QueryMetrics. Please note that this options will incurs overhead, so it should be
+     * enabled when debuging slow queries.
+     *
+     * @param indexMetricsPopulated
+     * @return indexMetricsPopulated
+     */
+    public CosmosQueryRequestOptions setIndexMetricsPopulated(boolean indexMetricsPopulated) {
```
[the remark this answers is not in the document]
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:509]: @simplynaveen20  - I suggested that name, as to be inline with `setQueryMetricsEnabled` and `isQueryMetricsEnabled` in CosmosQueryRequestOptions.java, and there are other booleans in that class as well, which have the same naming pattern.

**Comment by mbhaskar** [comment]
Any reason why we chose the tabular format?
- We want to move away from the existing tabular string printing format to a Json
- Only the client generated timeline metris like partition timeline / scheduling timeline etc.. were implemented in the text tabular format earlier which might eventually go away as we already have more granular and accurate client side diag.

I recommend printing these either in the format that is used by server side query metrics or as a simple json instead of the current way

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/WFConstants.java [hunk 6]
```diff
@@ -75,5 +75,6 @@
         public static final String ALLOW_TENTATIVE_WRITES = "x-ms-cosmos-allow-tentative-writes";
         public static final String IS_USER_REQUEST = "x-ms-cosmos-internal-is-user-request";
         public static final String BACKEND_REQUEST_DURATION_MILLISECONDS = "x-ms-request-duration-ms";
+        public static final String INDEX_UTILIZATION = "x-ms-cosmos-index-utilization";
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/WFConstants.java:78]: The same thing is done for x-ms-documentdb-query-metrics, which is in both the places HttpConstants.HttpHeaders and WFConstants.BackendHeaders. I followed the same pattern to be consistent.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    private String indexDocumentExpression;
+    @JsonProperty(value = "FilterPreciseSet", access = JsonProperty.Access.WRITE_ONLY)
+    private boolean filterExpressionPrecision;
+    @JsonProperty(value = "IndexPreciseSet", access = JsonProperty.Access.WRITE_ONLY)
+    private boolean indexPlanFullFidelity;
+    @JsonProperty(value = "IndexImpactScore", access = JsonProperty.Access.WRITE_ONLY)
+    private String indexImpactScore;
+
+    SingleIndexUtilizationEntity() {
+        super();
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java:22]: For Jackson to work on creating the object from JSON, we need a default constructor like this.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/IndexUtilizationInfo.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        return new IndexUtilizationInfo(
+            utilizedSingleIndexes,
+            potentialSingleIndexes,
+            utilizedCompositeIndexes,
+            potentialCompositeIndexes);
+    }
+
+    static IndexUtilizationInfo createFromJSONString(String jsonString) {
+        if (StringUtils.isEmpty(jsonString)) {
+            throw new NullPointerException("jsonString");
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/IndexUtilizationInfo.java:138]: Removed this check, as I am doing this check already in QueryMetrics.java (lines 290 - 292).

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java [hunk 9]
```diff
@@ -275,6 +287,10 @@ public static QueryMetrics createFromDelimitedStringAndClientSideMetrics(String
         double outputDocumentSize = metrics.get(QueryMetricsConstants.OutputDocumentSize);
         double retrievedDocumentSize = metrics.get(QueryMetricsConstants.RetrievedDocumentSize);
         Duration totalQueryExecutionTime = QueryMetricsUtils.getDurationFromMetrics(metrics, QueryMetricsConstants.TotalQueryExecutionTimeInMs);
+        IndexUtilizationInfo indexUtilizationInfo = null;
+        if (indexUtilizationInfoJSONString!=null) {
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java]: Fixed

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/IndexUtilizationInfo.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        }
+        ObjectMapper indexUtilizationInfoObjectMapper = new ObjectMapper();
+        IndexUtilizationInfo indexUtilizationInfo = null;
+        try {
+            indexUtilizationInfo = indexUtilizationInfoObjectMapper.readValue(jsonString, IndexUtilizationInfo.class);
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/IndexUtilizationInfo.java]: Added

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:509]: Enabled make sense but setIndexMetricsPopulated does not sound correct , i am just going with english

[the remark this answers is not in the document]
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:509]: @simplynaveen20 - how about we rename it to `indexMetricsEnabled` and getter and setter to be consistent with queryMetrics ?

like -> `isIndexMetricsEnabled()` and `setIndexMetricsEnabled()`?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SingleIndexUtilizationEntity.java]: Removed it

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java [hunk 11]
```diff
@@ -315,6 +315,48 @@ public void queryPlanDiagnostics() throws JsonProcessingException {
...
+            CosmosQueryRequestOptions options = new CosmosQueryRequestOptions();
+            options.setQueryMetricsEnabled(true);
+            options.setIndexMetricsPopulated(true);
+            Iterator<FeedResponse<InternalObjectNode>> iterator = cosmosContainer.queryItems(query, options, InternalObjectNode.class).iterableByPage().iterator();
+            while (iterator.hasNext()) {
+                FeedResponse<InternalObjectNode> feedResponse = iterator.next();
+                queryDiagnostics = feedResponse.getCosmosDiagnostics().toString();
+                logger.info("This is query diagnostics {}",queryDiagnostics);
+                if (feedResponse.getResponseHeaders().containsKey(HttpConstants.HttpHeaders.INDEX_UTILIZATION)) {
+                    assertThat(feedResponse.getResponseHeaders().get(HttpConstants.HttpHeaders.INDEX_UTILIZATION)).isNotNull();
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDiagnosticsTest.java:354]: Added the check.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java [hunk 12]
```diff
@@ -483,6 +485,34 @@ public CosmosQueryRequestOptions setThresholdForDiagnosticsOnTracer(Duration thr
         return this;
     }
 
+    /**
+     * Gets indexMetricsEnabled, which is used to obtain the index metrics to understand how the query engine used existing
+     * indexes and could use potential new indexes.
+     * The results will be displayed in QueryMetrics. Please note that this options will incurs overhead, so it should be
+     * enabled when debuging slow queries.
+     *
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:493]:
```suggestion
     * By default the indexMetrics are disabled.
```

**Review by kushagraThapar — APPROVED** [review]
LGTM with one minor suggestion. Thanks @aayush3011

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java (hunk 12 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:493]: I have that in line 494, to be consistent with isQueryMetricsEnabled().

**Comment by simplynaveen20** [comment]
As per the benchmark graph impact is > 2% , did we do multiple run to have more accurate result ?

**Comment by kushagraThapar** [comment]
> As per the benchmark graph impact is > 2% , did we do multiple run to have more accurate result ?

@simplynaveen20  - that was only 1 run, I am running the performance benchmark for multiple runs, will share the result soon.

**Pull request merged**